src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -78,6 +78,7 @@ ArchiveBuilder::SourceObjList::~SourceObjList() {
 void ArchiveBuilder::SourceObjList::append(SourceObjInfo* src_info) {
   // Save this source object for copying
   _objs->append(src_info);
+  src_info->set_index(_objs->length()-1);
```

Comment:
It's better to put this line before appending, so you can do

```
src_info->set_index(_objs->length());
```

That way you don't need to worry about ever getting a negative ID.